string configuration shortcuts

jmsdvl opened the pull request:
This is a first attempt at what implementing #15 might look like.  Any and all feedback is welcome.

jd commented on daiquiri/__init__.py:
```diff
@@ -88,8 +88,27 @@ def setup(level=logging.WARNING, outputs=[output.STDERR], program_name=None,
+        # instantiating an output from a dict of options
+        if isinstance(out, dict):
+            klassnames = out.keys()
+            for name in klassnames:
```
`for name in out:`

jd commented on daiquiri/__init__.py:
```diff
@@ -88,8 +88,27 @@ def setup(level=logging.WARNING, outputs=[output.STDERR], program_name=None,
+        if isinstance(out, dict):
+            klassnames = out.keys()
+            for name in klassnames:
+                klass = getattr(output, name)
```
This can raise `AttributeError`. Might be nice to catch and raise a locally defined exception.

jd commented on daiquiri/tests/test_daiquiri.py:
```diff
@@ -67,3 +67,12 @@ def test_set_default_log_levels(self):
     def test_parse_and_set_default_log_levels(self):
         daiquiri.parse_and_set_default_log_levels(
             ("urllib3=warn", "foobar=debug"))
+
```
You need to also test the `dict` version of output entry.

jd commented on daiquiri/__init__.py:
```diff
@@ -88,8 +88,27 @@ def setup(level=logging.WARNING, outputs=[output.STDERR], program_name=None,
-        o.add_to_logger(root_logger)
+    for out in outputs:
+        # instantiating an output from a dict of options
+        if isinstance(out, dict):
```
Not convinced by this dict approach. It seems this is just using
`{'syslog': {'facility': 'user1'}}`
Rather than
`output.Syslog(facility="user1")`

which to be honest is not a win. Maybe just ditch it entirely?

jmsdvl replied:
I'll leave that decision up to you.  I think it could be a win for people who want to read their conf from a JSON-ified source, but on the other hand, if they're doing that then probably they're talking to `logging` directly (since it can already do that iirc).

jd replied:
Ok, then leave it out of this PR. If anybody comes with good argument about that, we can revisit later in a new PR!

jmsdvl commented:
On further consideration I think you're correct, there's nothing to gain by using dicts for config - if there are user's who want to user JSON-like data for configuration, they most likely are also talking directly to the logging module.  Accordingly I've removed that option and updated the docs.

While we're on the topic though, should I add a few more basic names to the preconfigured outputs?  Something like `error.log` to create a file with loglevel error and a text formatter, or `info.json` to create a file with loglevel info and a json formatter?  Or anything else you like, really.

jd commented on daiquiri/__init__.py:
```diff
@@ -88,8 +88,20 @@ def setup(level=logging.WARNING, outputs=[output.STDERR], program_name=None,
-    for o in outputs:
-        o.add_to_logger(root_logger)
+    for out in outputs:
+
```
I'd go for a simpler version:

```python
for out in outputs:
    if isinstance(out, str):
        out = output.preconfigured.get(out)
        if not out:
            raise RuntimeError("Output `%s' is not available" % out)
    out.add_to_logger(root_logger)
```

And that's it!
